### PR TITLE
New attribute 'build_tags' for index_image operations

### PR DESCRIPTION
**Pull request by midnightercz**
For index image operations, it's now possible to set build_tags
attribute to apply extra tags for intermediate index image

### CODE REVIEW

**Comment by pbortlov** [comment]
Hi @midnightercz, what index-image operations do you mean? Add bundles, Rm Operator, Merge index image?

**Comment by midnightercz** [comment]
@pbortlov I meant add and remove. But it's true that it should be probably also added to merge index image as well

**Comment by pbortlov** [comment]
Could we wait for #229 to be merged to add this variable to the new endpoint?

**Comment by midnightercz** [comment]
@pbortlov yeah, we can wait

File: iib/web/models.py
Comment (yashvardhannanavati) [inline iib/web/models.py:1033]: I don't think you need this validation here right? Since you are calling the parent class `_from_json`?

File: iib/web/models.py
Comment (yashvardhannanavati) [inline iib/web/models.py:1057]: I would prefer having this logic in the `api_v1.yaml` for all the requests.

Comment (yashvardhannanavati) [reply iib/web/models.py:1057]: Looks like similar code to me for all requests. Maybe you can create a helper method in ` api_v1.yaml` ?

File: iib/web/models.py
Comment (yashvardhannanavati) [inline iib/web/models.py:1325]: Same as above

File: iib/workers/tasks/build.py
Comment (yashvardhannanavati) [inline iib/workers/tasks/build.py:1010]: could you please add a comment here as to why we are using the first element of the list?

File: iib/workers/tasks/build_regenerate_bundle.py
Comment (yashvardhannanavati) [inline iib/workers/tasks/build_regenerate_bundle.py:121]: You are accepting `build_tags` from the user in the API and not using them at all. I think we shouldn't give an option at all, no?
Am I missing something?

**Review by yashvardhannanavati — CHANGES_REQUESTED** [review]

File: iib/workers/tasks/build.py
Comment (pbortlov) [inline iib/workers/tasks/build.py:132]: Why is this removed?

File: iib/workers/tasks/build.py
Comment (pbortlov) [inline iib/workers/tasks/build.py:130]: if I understand correctly, every pull spec will be appended with each arch?
Could you give me an example how the yaml file will look like in the end?

File: iib/workers/tasks/build_merge_index_image.py
Comment (pbortlov) [inline iib/workers/tasks/build_merge_index_image.py:138]: why is here an empty list?

File: tests/test_workers/test_tasks/test_build_regenerate_bundle.py
Comment (pbortlov) [inline tests/test_workers/test_tasks/test_build_regenerate_bundle.py:118]: I am missing tests for create_empty_index, merge_index_image

**Review by pbortlov — CHANGES_REQUESTED** [review]

File: iib/web/models.py
Comment (midnightercz) [reply iib/web/models.py:1033]: but there's no validation for build_tags there

File: iib/web/models.py
Comment (midnightercz) [reply iib/web/models.py:1057]: I'm not sure I understand your request. Logic in yaml?

File: iib/workers/tasks/build.py
Comment (midnightercz) [reply iib/workers/tasks/build.py:132]: this was just moved bellow

File: iib/workers/tasks/build.py
Comment (midnightercz) [reply iib/workers/tasks/build.py:130]: that's correct. Yaml files will look very same as they were before this change

File: iib/workers/tasks/build_merge_index_image.py
Comment (midnightercz) [reply iib/workers/tasks/build_merge_index_image.py:138]: because it's not needed here. Final call of _create_and_push_manifest_list is in handle_merge_request

File: iib/workers/tasks/build_regenerate_bundle.py
Comment (midnightercz) [reply iib/workers/tasks/build_regenerate_bundle.py:121]: it's shared method, so if you don't want to use them, you need to pass there empty list. Build_tags are not accepted for regenerate_bundles. Or do you think we should support it as well?

File: tests/test_workers/test_tasks/test_build_regenerate_bundle.py
Comment (midnightercz) [reply tests/test_workers/test_tasks/test_build_regenerate_bundle.py:118]: because there weren't any which works with real _create_and_push_manifest_list

File: iib/workers/tasks/build.py
Comment (pbortlov) [reply iib/workers/tasks/build.py:130]: thanks. I saw the example in tests

File: tests/test_workers/test_tasks/test_build_regenerate_bundle.py
Comment (pbortlov) [reply tests/test_workers/test_tasks/test_build_regenerate_bundle.py:118]: oh, that's why. Thank you for explaining :)

File: iib/web/models.py
Comment (pbortlov) [reply iib/web/models.py:658]: I think so, because `get_request_query_options` is called by `get_builds`. Could you create tasks for both of them?

Comment (pbortlov) [reply iib/web/models.py:658]: THanks for pointing this out @lipoja

File: iib/web/models.py
Comment (pbortlov) [reply iib/web/models.py:1034]: Do you need braces around the build_tags?

Comment (pbortlov) [reply iib/web/models.py:1034]: I will add it to the Request._from_json if you want to use it for RequestAdd, RequestRm, RequestCreateEmptyImage,

File: iib/web/models.py
Comment (pbortlov) [inline iib/web/models.py:1063]: What about request_kwargs.pop('build_tags', []) instead of those 3 lines?
This logic can be done in Request._from_json() to avoid uplicates

File: iib/web/models.py
Comment (pbortlov) [inline iib/web/models.py:958]: What about request_kwargs.pop('build_tags', []) instead of those 3 lines?
This logic can be done in Request._from_json() to avoid uplicates

File: iib/web/models.py
Comment (pbortlov) [inline iib/web/models.py:914]: Let's check this in Request._from_json

File: iib/web/models.py
Comment (pbortlov) [inline iib/web/models.py:1329]: Please add it to the Request

File: iib/web/models.py
Comment (pbortlov) [inline iib/web/models.py:1363]: Interestingly, you didn't add this line to the classes above RequestAdd and RequestRm. What about to add it to the Request class and this line will ne inherited for RequestAdd, RequestRm, RequestRegenerateBundle and RequestCreateEmptyImage.

File: iib/workers/tasks/build.py
Comment (pbortlov) [inline iib/workers/tasks/build.py:1046]: Should it here None instead?

**Review by pbortlov — CHANGES_REQUESTED** [review]

File: iib/web/models.py
Comment (midnightercz) [reply iib/web/models.py:1063]: yeah, this can be reduced into one line

File: iib/web/models.py
Comment (midnightercz) [reply iib/web/models.py:914]: I'm not sure build_tags should be available for all requests. Therefore I added it per class

File: iib/web/models.py
Comment (midnightercz) [reply iib/web/models.py:1363]: I added it here only because of the inheritance you mentioned.

File: iib/workers/tasks/build.py
Comment (midnightercz) [reply iib/workers/tasks/build.py:1046]: it doesn't matter if it's [] or None. But yeah, I should have been consistent

File: iib/web/models.py
Comment (yashvardhannanavati) [reply iib/web/models.py:1058]: Sorry I meant api_v1.py

File: iib/web/models.py
Comment (yashvardhannanavati) [reply iib/web/models.py:914]: Since you have added it under additional_optional_params in `_from_json`, let's have this check in there if the user has provided `build_tags`. That would mean removing it from here.

File: iib/web/models.py
Comment (yashvardhannanavati) [reply iib/web/models.py:1034]: Same as above, moving this check from here to _from_json will help you get rid of this code duplication.

Comment (yashvardhannanavati) [reply iib/web/models.py:1034]: Simply pass `build_tags` in the call on line 1050 while calling `cls._from_json` under `additional_optional_params ` and that would do the trick.

Comment (yashvardhannanavati) [reply iib/web/models.py:1034]: Same can be done for all three classes `RequestAdd`, `RequestRm` and `RequestMergeIndex`

File: iib/web/models.py
Comment (yashvardhannanavati) [reply iib/web/models.py:1058]: So for all the three types of requests, let's have this piece of code in `api_v1.py` file instead.

File: iib/web/models.py
Comment (yashvardhannanavati) [reply iib/web/models.py:317]: Ah, I see it the merge index isn't inheriting mixin class. In that case, let's add this in Mixin for `Add` and `Rm` and in the `RequestMergeIndex` class too. That will solve your problem

File: iib/web/models.py
Comment (yashvardhannanavati) [reply iib/web/models.py:1324]: Let's move this to `api_v1.py` and you will have to duplicate the validation here since you are not calling `cls._from_json` here since the Mixin class is not being inherited here.

File: iib/workers/tasks/build_regenerate_bundle.py
Comment (yashvardhannanavati) [reply iib/workers/tasks/build_regenerate_bundle.py:121]: No, I agree they shouldn't be supported in regenerate-bundle.

**Review by yashvardhannanavati — CHANGES_REQUESTED** [review]
@midnightercz This is looking good except for the changes I suggested in `models.py`. Please feel free to setup a call if my comments aren't clear.

File: iib/web/models.py
Comment (midnightercz) [reply iib/web/models.py:317]: That's not the only reason. Another reason is back_populates attribute which points to Request model. I'm not sure if it's still going to work if I specify that in RequestRm but attribute itself won't be in Request. Also I don't think you can set multiple values in back_populates.
Therefore I simply set build_tags to None for classes where it's not relevant

File: iib/workers/tasks/build.py
Comment (pbortlov) [inline iib/workers/tasks/build.py:119]: Typo: 'intermetdiate'

**Comment by pbortlov** [comment]
hi @midnightercz , could you please summarize where should be the build_tags? I already got confused about where the attribute is needed and why?

File: iib/web/models.py
Comment (yashvardhannanavati) [reply iib/web/models.py:1042]: It's not required because it's already present in `_from_json`

File: iib/web/models.py
Comment (yashvardhannanavati) [inline iib/web/models.py:963]: You have a function to do this. `add_build_tag`

File: iib/web/models.py
Comment (yashvardhannanavati) [inline iib/web/models.py:1051]: Same, you have a function for this. `add_build_tag`

File: iib/web/models.py
Comment (yashvardhannanavati) [inline iib/web/models.py:1107]: If it's present in `Request`, you can simply set `nullable=True` in Request and this won't be required.

File: iib/web/models.py
Comment (yashvardhannanavati) [inline iib/web/models.py:1323]: Same as above.

**Review by yashvardhannanavati — CHANGES_REQUESTED** [review]

File: iib/web/models.py
Comment (midnightercz) [reply iib/web/models.py:963]: yes, that's why I'm using it. Or I don't get your point

File: iib/web/models.py
Comment (midnightercz) [reply iib/web/models.py:1051]: hmm, I don't understand

File: iib/web/models.py
Comment (midnightercz) [reply iib/web/models.py:1107]: According to docs sqlalchemy.orm.relationship doesn't accept nullable attribute. Or do you mean something else?

File: iib/web/models.py
Comment (midnightercz) [reply iib/web/models.py:958]: fixed

File: iib/web/models.py
Comment (midnightercz) [reply iib/web/models.py]: fixed

File: iib/workers/tasks/build.py
Comment (midnightercz) [reply iib/workers/tasks/build.py]: fixed

File: iib/web/models.py
Comment (yashvardhannanavati) [reply iib/web/models.py:961]: Sorry I think I misunderstood something here.

File: iib/web/models.py
Comment (yashvardhannanavati) [inline iib/web/models.py:1040]: If you remove the last comma, it will go back to as it was before. I'm guessing this change was unintentional and Python black formatting.

File: iib/workers/tasks/build.py
Comment (yashvardhannanavati) [inline iib/workers/tasks/build.py:109]: This will have to be refactored as a result of https://github.com/release-engineering/iib/pull/286

File: iib/workers/tasks/build.py
Comment (yashvardhannanavati) [inline iib/workers/tasks/build.py:853]: I think something got messed up in the rebase. Could you please fix it

**Review by yashvardhannanavati — CHANGES_REQUESTED** [review]
Very close to merging. Could you please refactor the `_create_and_push_manifest_list` function and fix the rebase error.

File: iib/workers/tasks/build.py
Comment (midnightercz) [reply iib/workers/tasks/build.py:853]: sure

File: iib/workers/tasks/build.py
Comment (midnightercz) [reply iib/workers/tasks/build.py:155]: that's correct

**Review by pbortlov — APPROVED** [review]

**Review by lipoja — APPROVED** [review]

File: iib/workers/tasks/build.py
Comment (yashvardhannanavati) [inline iib/workers/tasks/build.py:129]: Why does this for loop create a new manifest list for every tag? Can't we simply put this for loop just before line 149? That way we simply push the same manifest list with different tags.

Comment (midnightercz) [reply iib/workers/tasks/build.py:129]: yeah, I can improve that

**Review by pbortlov — DISMISSED** [review]

File: iib/web/models.py
Comment (yashvardhannanavati) [inline iib/web/models.py:244]: I'm sorry I missed this earlier. But could you please use an association table like`RequestArchitecture` to be consistent. Something like `RequestBuildTag`

File: iib/web/models.py
Comment (yashvardhannanavati) [inline iib/web/models.py:373]: Also, shouldn't we be checking if a build tag already exists before adding a new one associated with the request? Check out `add_architecture` for inspiration.

Comment (midnightercz) [reply iib/web/models.py:373]: Yeah, we should. I thought sqlalchemy is smart enough to figure this out

File: iib/web/models.py
Comment (yashvardhannanavati) [inline iib/web/models.py:384]: This should probably be `if bt not in self.build_tags`

File: iib/web/models.py
Comment (yashvardhannanavati) [inline iib/web/models.py:244]: @midnightercz  I believe we talked about creating an association table here. `RequestBuildTag` should only be associating the two tables. Like `RequestArchitecture` on line 158. 

Another table called the `BuildTag` should house the `id` and `name` like the `Architecture` table. Could you please change it?

**Review by yashvardhannanavati — CHANGES_REQUESTED** [review]

File: iib/web/models.py
Comment (midnightercz) [reply iib/web/models.py:384]: no, it's not needed if you set request_id=self.id. You can check newly added tests.

File: iib/web/models.py
Comment (midnightercz) [reply iib/web/models.py:244]: I get your point, but unlike architecture which can be chosen only from limited set of choices, build tag can be any string and most probably even will be unique for every request. Does it make sense then to create enumeration table and association table?

Comment (yashvardhannanavati) [reply iib/web/models.py:244]: It would be nice to add a `__repr__` function too.

File: iib/web/models.py
Comment (pbortlov) [inline iib/web/models.py:252]: This line is the same as on line no. 245. Is it really the same?

File: iib/workers/tasks/build.py
Comment (pbortlov) [inline iib/workers/tasks/build.py:155]: Could you add a comment on why is taken the first item of the list?

File: iib/workers/tasks/build.py
Comment (pbortlov) [inline iib/workers/tasks/build.py:988]: Is it singular or plural? Do we get the one pull spec or more? I guess one but better to ask

File: iib/workers/tasks/build_merge_index_image.py
Comment (pbortlov) [inline iib/workers/tasks/build_merge_index_image.py:278]: Here is plural and in the file `build.py` is just `output_pull_spec` instead of `spec**s**`

Comment (pbortlov) [reply iib/workers/tasks/build_merge_index_image.py:278]: it seems there is a discrepancy. THe function `_create_and_push_manifest_list` returns `return output_pull_specs[0]`. You are using it differently here.

File: iib/web/models.py
Comment (midnightercz) [reply iib/web/models.py:252]: right, I'll fix the description

File: iib/workers/tasks/build.py
Comment (midnightercz) [reply iib/workers/tasks/build.py:155]: You mean in the comment or here? First item is taken as first item is production tag, next items in the list are extra build tags

Comment (pbortlov) [reply iib/workers/tasks/build.py:155]: Comment in code :)

File: iib/workers/tasks/build_create_empty_index.py
Comment (midnightercz) [reply iib/workers/tasks/build_create_empty_index.py]: removed

File: tests/test_web/test_api_v1.py
Comment (midnightercz) [reply tests/test_web/test_api_v1.py]: Yeah, it's fro generic request schema. This is why I didn't want to add it to generic _from_json which Yash insisted on.

**Review by yashvardhannanavati — DISMISSED** [review]
We just need to confirm that this has been tested on the local dev env.

**Comment by yashvardhannanavati** [comment]
@midnightercz Could you please rebase this?

**Comment by yashvardhannanavati** [comment]
@midnightercz Could you please squash your commits :)

File: iib/workers/tasks/build.py
Comment (pbortlov) [inline iib/workers/tasks/build.py:129]: Could you please add a comment here on why you iterate on  output_pull_spec and rewriting it?

**Review by yashvardhannanavati — DISMISSED** [review]

**Review by pbortlov — APPROVED** [review]

**Pull request merged**
